ci: build generic status-go and all shells

jakubgs opened the pull request:
When discussing caching of `status-go` with Sid I noticed that the build we cache daily created from our nightly job is different from the build we create locally due to a single input.

In a release CI host we can see the IPFS URL is that of Infura:
```
 > find /nix/store -maxdepth 1 -name '*-status-go-*android' | tail -n1
/nix/store/2cc8ilhx5g3k2awbn4sla61n4cml2405-status-go-0.130.1-d2cce5e-android

 > RESULT=$(find /nix/store -maxdepth 1 -name '*-status-go-*android' | tail -n1)

 > nix show-derivation $RESULT | tr ' ' '\n' | grep IpfsGateway
github.com/status-im/status-go/params.IpfsGatewayURL=https://status-im.infura-ipfs.io/ipfs/
```
But for a local build the URL is the default, which is our own gateway:
```
 > nix-build --no-out-link -A targets.status-go.mobile.android
/nix/store/1p53m7a6y1kg3vcyd8d06scf3bsyn5rk-status-go-0.157.2-47711c4-android

 > RESULT=$(nix-build --no-out-link -A targets.status-go.mobile.android)

 > nix show-derivation $RESULT | tr ' ' '\n' | grep IpfsGateway
github.com/status-im/status-go/params.IpfsGatewayURL=https://ipfs.status.im/
```
This difference causes builds of `status-go` that get uploaded to our Nix cache to not match what developers locally would build, which results in a cache miss.

This changes the Nix cache CI jobs to instead of building only dependencies(`buildInuts`) to simply build the generic versions of `status-go` without nightly specific inputs.

This also fixes configuration of our own Nix cache:

According to this line from the docs:

>The system-wide configuration file sysconfdir/nix/nix.conf (i.e. /etc/nix/nix.conf),
>or $NIX_CONF_DIR/nix.conf if NIX_CONF_DIR is set. Values loaded in this file are not
>forwarded to the Nix daemon. The client assumes that the daemon has already loaded them.

https://nixos.org/manual/nix/stable/command-ref/conf-file.html#description

Our usage of `NIX_CONF_DIR` has been wrong for a while now.
The correct way of applying this config is using `NIX_USER_CONF_FILES`.

In addition the `extra-substituters` no longer exists in the docs.
Use of `trusted-substituters` is necessary according to:

>At least one of the following conditions must be met for Nix to use a substituter:
>
>- the substituter is in the trusted-substituters list
>- the user calling Nix is in the trusted-users list

https://nixos.org/manual/nix/stable/command-ref/conf-file.html#conf-substituters

jakubgs committed:
ci: build generic status-go and all shells

When discussing caching of `status-go` with Sid I noticed that the build
we cache daily created from our nightly build is different from the
build we create locally due to a single input.

In a release CI host we can see the IPFS URL is that of Infura:
```
 > find /nix/store -maxdepth 1 -name '*-status-go-*android' | tail -n1
/nix/store/2cc8ilhx5g3k2awbn4sla61n4cml2405-status-go-0.130.1-d2cce5e-android

 > RESULT=$(find /nix/store -maxdepth 1 -name '*-status-go-*android' | tail -n1)

 > nix show-derivation $RESULT | tr ' ' '\n' | grep IpfsGateway
github.com/status-im/status-go/params.IpfsGatewayURL=https://status-im.infura-ipfs.io/ipfs/
```
But for a local build the URL is the default, which is our own gateway:
```
 > nix-build --no-out-link -A targets.status-go.mobile.android
/nix/store/1p53m7a6y1kg3vcyd8d06scf3bsyn5rk-status-go-0.157.2-47711c4-android

 > RESULT=$(nix-build --no-out-link -A targets.status-go.mobile.android)

 > nix show-derivation $RESULT | tr ' ' '\n' | grep IpfsGateway
github.com/status-im/status-go/params.IpfsGatewayURL=https://ipfs.status.im/
```
This difference causes builds of `status-go` that get uploaded to our
Nix cache to not match what developers locally would build, which
results in a cache miss.

This changes the Nix cache CI jobs to instead of building only dependencies
(`buildInuts`) to simply build the generic versions of `status-go` without
nightly specific inputs.

Signed-off-by: Jakub Sokołowski <jakub@status.im>

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | c2ff7ed2 | [#1](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-16287/1/) | 2023-06-15 07:58:57 | ~5 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-230615-075313-c2ff7e-pr16287-x86.apk) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-230615-075313-c2ff7e-pr16287-x86.apk)|

jakubgs commented:
This is something worth nothing from Nix docs:

>By default Nix reads settings from the following places:
>
>* The system-wide configuration file sysconfdir/nix/nix.conf (i.e. /etc/nix/nix.conf on most systems), or $NIX_CONF_DIR/nix.conf if NIX_CONF_DIR is set. __Values loaded in this file are not forwarded to the Nix daemon. The client assumes that the daemon has already loaded them.__
>* If NIX_USER_CONF_FILES is set, then each path separated by : will be loaded in reverse order.
>* Otherwise it will look for nix/nix.conf files in XDG_CONFIG_DIRS and XDG_CONFIG_HOME. If unset, XDG_CONFIG_DIRS defaults to /etc/xdg, and XDG_CONFIG_HOME defaults to $HOME/.config as per [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html).
>* If NIX_CONFIG is set, its contents is treated as the contents of a configuration file.

https://nixos.org/manual/nix/stable/command-ref/conf-file.html#description

The key part is here:

>Values loaded in this file are not forwarded to the Nix daemon.
>The client assumes that the daemon has already loaded them.

This means that on NixOS or a in a multi-user setup this configuration dot not have any effect.

jakubgs committed:
nix: fix applying nix.conf by using correct env var

According to this line from the docs:

>The system-wide configuration file sysconfdir/nix/nix.conf (i.e. /etc/nix/nix.conf),
>or $NIX_CONF_DIR/nix.conf if NIX_CONF_DIR is set. Values loaded in this file are not
>forwarded to the Nix daemon. The client assumes that the daemon has already loaded them.

https://nixos.org/manual/nix/stable/command-ref/conf-file.html#description

Our usage of `NIX_CONF_DIR` has been wrong for a while now.
The correct way of applying this config is using `NIX_USER_CONF_FILES`.

In addition the `extra-substituters` no longer exists in the docs.
Use of `trusted-substituters` is necessary according to:

>At least one of the following conditions must be met for Nix to use a substituter:
>
>- the substituter is in the trusted-substituters list
>- the user calling Nix is in the trusted-users list

https://nixos.org/manual/nix/stable/command-ref/conf-file.html#conf-substituters

Signed-off-by: Jakub Sokołowski <jakub@status.im>

jakubgs commented:
Looks like this needs some work as two devs reported seeing this warning:
```
warning: ignoring untrusted substituter 'https://nix-cache.status.im/'
```

jakubgs commented:
Our settings are this:
https://github.com/status-im/status-mobile/blob/8ad04a28a5180e2302f75ee87effb4d7320b3002/nix/nix.conf#L2-L4
Which clearly sets [`trusted-substituters`](https://nixos.org/manual/nix/stable/command-ref/conf-file.html#conf-trusted-substituters), and that config is clearly loaded:
```
 > make shell
Configuring Nix shell for target 'default'...

[nix-shell:~/work/status-mobile]$ nix show-config | grep substituters
substituters = https://nix-cache.status.im/ https://cache.nixos.org/
trusted-substituters = https://nix-cache.status.im/
```
So something else is wrong.

jakubgs commented:
Now that the Nix upgrade to `2.14.1` has been merged I can go forward with fixing this caching configuration:

* https://github.com/status-im/status-mobile/pull/16297

jakubgs commented:
After upgrading to Nix `2.14.1` I have tested this on a Linux CI host and it appears to work:
```
admin@linux-01.he-eu-hel1.ci.devel:~/status-mobile % make status-go-android           
nix/scripts/build.sh targets.status-go.mobile.android
Running: nix-build --pure --fallback --no-out-link --show-trace --attr targets.status-go.mobile.android 
this path will be fetched (0.00 MiB download, 73.73 MiB unpacked):
  /nix/store/1p53m7a6y1kg3vcyd8d06scf3bsyn5rk-status-go-0.157.2-47711c4-android
copying path '/nix/store/1p53m7a6y1kg3vcyd8d06scf3bsyn5rk-status-go-0.157.2-47711c4-android' from 'https://nix-cache.status.im'...

Extracting result: /nix/store/1p53m7a6y1kg3vcyd8d06scf3bsyn5rk-status-go-0.157.2-47711c4-android
'/nix/store/1p53m7a6y1kg3vcyd8d06scf3bsyn5rk-status-go-0.157.2-47711c4-android/status-go-47711c4.aar' -> './result/status-go-47711c4.aar'

SUCCESS
```
But on MacOS it fails:
```
admin@macm1-01.ih-eu-mda1.ci.devel:~/status-mobile % make status-go-android
nix/scripts/build.sh targets.status-go.mobile.android
warning: ignoring untrusted substituter 'https://nix-cache.status.im/', you are not a trusted user.
...
```
So something is different there.

jakubgs commented:
And I guess this is the main difference:
```
admin@linux-01.he-eu-hel1.ci.devel:~ % sudo grep trusted-users /etc/nix/nix.conf
trusted-users = jenkins @adm
```
```
admin@macm1-01.ih-eu-mda1.ci.devel:~/status-mobile % sudo grep trusted-users /etc/nix/nix.conf 
trusted-users = jenkins
```

jakubgs commented:
And indeed, adding `admin` group to `trusted-users` does fix it in this use case:
```
admin@macm1-01.ih-eu-mda1.ci.devel:~/status-mobile % make status-go-android
nix/scripts/build.sh targets.status-go.mobile.android
Running: nix-build --pure --fallback --no-out-link --show-trace --attr targets.status-go.mobile.android 
these 2 derivations will be built:
  /nix/store/ciq59ddmsrs8a59pxhhaqscgmv7rcpc7-gomobile-unstable-2022-05-18.drv
  /nix/store/ikd6sdwcd8hxld9ij5mfd9kir42cngcn-status-go-0.157.2-47711c4-android.drv
...
```
But it doesn't see to have been cached.
